proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,477 @@
+  - Verify the program account header
+  - Check that the status stored in the program account is `Deployed`,
+    otherwise throw `InvalidArgument`
+  - Check that the slot stored in the program account is not the current
+    (deployment cooldown), otherwise throw `InvalidArgument`
+  - Note: The slot is **not** set to the current slot to allow a
+  retract-modify-redeploy-sequence within the same slot or even within the
+  same transaction.
+  - Change the status stored in the program account to `Retracted`
+  - Set the `is_executable` flag to `false`
+
+#### Initialize
+
+- Instruction accounts:
+  - `[writable]` The account to initialize as program account.
+  - `[signer]` The new authority of the program.
+- Instruction data:
+  - Enum variant `7u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that the owner of the program account is loader-v4,
+    otherwise throw `InvalidAccountOwner`
+  - Check that the program account is writable,
+    otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the header,
+    otherwise throw `InsufficientFunds`
+  - Check that the new authority (instruction account at index 1) has signed,
+    otherwise throw `MissingRequiredSignature`
+  - Set the length of the program account to the header size
+  - Set the slot to zero, **not** the current slot
+  - Change the status stored in the program account to `NeverBeenDeployed`
+  - Copy the new authority address into the program account
+
+#### SetAuthority
+
+- Instruction accounts:
+  - `[writable]` The program account to change the authority of.
+  - `[signer]` The current authority of the program.
+  - `[signer]` The new authority of the program.
+- Instruction data:
+  - Enum variant `8u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the new authority (instruction account at index 2) has signed,
+    otherwise throw `MissingRequiredSignature`
+  - Check that the current authority is different from the new authority,
+    otherwise throw `InvalidArgument`
+  - Copy the new authority address into the program account
+
+#### Finalize
+
+- Instruction accounts:
+  - `[writable]` The program account to finalize.
+  - `[signer]` The current authority of the program.
+- Instruction data:
+  - Enum variant `9u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account header
+  - Check that the status stored in the program account is `Deployed` or
+    that the status is `Retracted` and the program length is 0 (header only),
+    otherwise throw `InvalidArgument`
+  - Change the status stored in the program account to `Finalized`
```

Comment:
What do you think about adding an additional `RetractedFinalized` status too? The current design of `Finalize` makes a finalized `Retracted` program look kinda executable from looking at the status alone. It requires the runtime to verify the program (ie check the program length is 0) to know that it's actually retracted.

Reply:
> It requires the runtime to verify the program (ie check the program length is 0) to know that it's actually retracted.

That is only once in the `Finalize` instruction and even if we added an additional enum item this would remain the case because the `Finalize` instruction is what sets that status. `Finalized` and `RetractedFinalized` would behave exactly the same in all loader instructions and you always have to load the account data anyway to see the status, thus you always know the length too. I don't really see the benefit of distinguishing between these two specifically yet.

Reply:
Yeah fair point about needing to load the account data to see the status anyways. I'm fine with keeping it as a single `Finalized` status and distinguishing "retracted" by checking program data length is zero